Improve visual organization of examples

colleenmcginnis opened the pull request:
Proposes a new system for organizing the list of Mapbox GL JS examples based on discussion in #2242 . Open to any suggestions on category names, order, which category a particular example falls in, and structure using jekyll. *I would also like to expand the search to include common keywords (not just titles) in future iterations.* 

![image](https://cloud.githubusercontent.com/assets/10479155/18492444/3aa94b86-79c0-11e6-972b-0105bc85fdae.png)

**Changes**: 
1.  Search bar at the top filters by example title
2.  Examples are organized into categories/subcategories (full list of examples in each category/subcategory [here](https://gist.github.com/colleenmcginnis/917a5d543e800d1fd1c5d4fee235f9c1)):
  - Getting started
    - Add a style
    - Add a layer
  - Dynamic styling
    - Style points
    - Style lines
    - Style polygons
  - User interaction
    - Popups
    - Filter
  - Animations
  - Plugins
    - Geocoding
    - Directions
    - GL Draw
  - Change map options
  - Misc

Changes to **_layouts** and **_posts** were modeled after the equivalent files for [mapbox.js](https://github.com/mapbox/mapbox.js/blob/mb-pages/docs/_layouts/pages.html). The **pages** layout now has all example categories/subcategories in the front matter (like in the page for mapbox.js examples), but I'm open to suggestions to make this easier to work with! Each example has tags in the front matter, like this: 
```
---
layout: example
category: example
title: Display a map
description: Initialize a map in an HTML element with Mapbox GL JS.
permalink: /examples/
tags:
    - get-started
---
```

## Checklist
Review by @mollymerp @lucaswoj @mapbox/support, specifically looking for feedback on: 
 - [ ] categories/subcategories
 - [ ] assignment of individual examples 
 - [ ] use of jekyll
 - [ ] general feedback on visual organization

peterqliu commented:
❤️  long overdue. thanks for jumping on this @colleenmcginnis

geografa commented:
This is great and categories look 👌 . Looking at [iOS examples](https://www.mapbox.com/ios-sdk/examples/) and [Mapbox.js examples](https://www.mapbox.com/mapbox.js/examples) pages, these seem consistent.

colleenmcginnis commented:
@lucaswoj this happened to @mollymerp too. She had to update to the most recent version of jekyll.

lucaswoj commented:
Got it. Thanks @colleenmcginnis. Updated the docs (and my local Jekyll version) in https://github.com/mapbox/mapbox-gl-js/pull/3181/commits/849f93d6101ae6ab545c457d0f996dd4a922bdf6

lucaswoj commented on docs/_layouts/pages.html:
```diff
@@ -2,25 +2,106 @@
+          {% for example in layout.examples %}
+          <div class='space-bottom1'>
+            <{{example.level}} class='heading {{example.class}}'>{{example.name}}</{{example.level}}>
+            {% for listing in site.tags[example.id] order:ascending %}
```
This line is triggering build errors for me

```
Liquid Warning: Liquid syntax error (line 32): Invalid attribute in for loop. Valid attributes are limit and offset in "listing in site.tags[example.id] order:ascending" in /_layouts/pages.html
```

The [Liquid docs](https://help.shopify.com/themes/liquid/tags/iteration-tags) do not list `order` as a valid attribute

colleenmcginnis replied:
Removed `order`. Just out of curiosity where did you find that error?

lucaswoj replied:
This error was in the console where I was running `jekyll`

lucaswoj commented on docs/_layouts/pages.html:
```diff
@@ -2,25 +2,106 @@
+    level: h3
+  - name: Misc
+    id: misc
+    level: h3
```
It's challenging to fit all our examples into this taxonomy.

 - "Getting started" suggests that it contains only the most basic examples but it includes some advanced examples like "Add live realtime data". 
 - There exists a "Dynamic styling" section but almost all examples in other sections use the runtime styling API. 
 - "Move symbol with the keyboard" is under "Animation" but there exists a "User Interaction" section. 

What do you think about this taxonomy, inspired by the GL JS docs and style spec?

 - Getting Started
 - Sources
 - Styling
 - Controls and Overlays
 - Interaction Handlers
 - Plugins

colleenmcginnis replied:
Did some renaming of categories and rearranging of examples. Take a look and let me know what you think!

lucaswoj commented on docs/_layouts/pages.html:
```diff
@@ -70,3 +151,54 @@
+    }
+
+
+ 
```
We don't use the double (or triple) newline formatting pattern in the GL JS codebase.

colleenmcginnis replied:
Changed when moving to site.js.

lucaswoj commented on docs/_layouts/pages.html:
```diff
@@ -70,3 +151,54 @@
+    
+
+  }); 
+</script>
```
Can you put this code in a separate `*.js` file so that the linter will run on it?

colleenmcginnis replied:
Done!

lucaswoj replied:
I don't see any changes to `site.js` and the search box has stopped working. Did you forget to commit the file?

colleenmcginnis commented:
Ready for your 👀 again @lucaswoj. 

Changes in this iteration: 
- Matched categories to the list you suggested 👍 
- Shifted some examples around (with input from @mollymerp)
- Moved js code to site.js
- Removed `order` from the `for` loop
- Added the new *Navigate the map with game-like controls* example

lucaswoj commented:
Just pushed some taxonomy rejiggering in https://github.com/mapbox/mapbox-gl-js/pull/3181/commits/38447a89005154767911edaee9de645f9e3b9087. Let me know if you want to discuss anything in there.

lucaswoj commented:
https://github.com/mapbox/mapbox-gl-js/pull/3181#discussion_r79916904 is the only thing blocking 🚢ing this

mollymerp commented:
@colleenmcginnis just pulled and the filter seems to be working properly now 🎈 

@lucaswoj I prefer the earlier version with the subcategories (https://github.com/mapbox/mapbox-gl-js/pull/3181/commits/de7dbddfe9ad1c438a50ac0cc8fa8e34bbea61b5). It's easier to parse visually and orient yourself and I don't think these new categories offer enough specificity to be particularly useful. What was your reasoning behind removing them?

lucaswoj commented:
>  I prefer the earlier version with the subcategories

No objection to subcatories per se. After reorganizing the examples there was only one subcategory left so I dropped the feature.  

>  I don't think these new categories offer enough specificity to be particularly useful. What was your reasoning behind removing them?

The new categories closely mirror the structure of our existing API documentation. 

The old categories 

 - in some cases did not use proper GL terminology (i.e "Moving around the map" vs "Camera" (see below))
 - in some cases were too specific to accurately describe the examples they contained (i.e. "Add a style" contained "Display a satellite map"  which doesn't "add" a style, it initializes the map with a style)
 - in some cases were too vague to justify not putting more examples in them (i.e. Why doesn't "Show and hide layers" go into "Dynamic styling"?)
 - in some cases overlapped (i.e. "Add a layer" vs "Dynamic styling")

> I don't think `Camera` is a useful category for users who are not familiar w/ GL terminology...

This is the documentation where users become familiar with GL terminology. Camera is standard terminology for any rendering software.

mollymerp commented:
I think the API docs and the examples have slightly different audiences and shouldn't necessarily have the same categories. "Getting started" and "moving around the map" are approachable and anyone would be able to understand their meaning without spending time acquainting themselves with the full GL API. ¯\_(ツ)_/¯ 

@colleenmcginnis 🚢  if you're ready!

danswick commented:
So glad you took this on, @colleenmcginnis! It's been a long time coming. 

I agree with @mollymerp and @colleenmcginnis that the audience for the examples page could be better served by more approachable categories. The sub-categories in Colleen's original commit also help explain some of the underlying concepts behind GL JS and the Style Spec and do a nice job of chunking similar concpets. 

We can always ship a version of this and iterate later, but I took some of the original categories / subcategories and tweaked them a little bit to address some of @lucaswoj's comments while retaining much of the original structure: 

---

## Getting started

- Display a map  

### Set a style

- Display a map with a custom style 
- Display a satellite map   

## Sources & layers

- Add a vector tile source
- Add a third party vector tile source  
- Add a raster tile source
- Add a new layer below labels  

### Point data

- Draw GeoJSON points 
- Add custom icons with Markers  

### Clustered point data
- Create a heatmap from points  
- Create and style clusters  

### Line data

- Add a GeoJSON line

### Polygon data

- Add a GeoJSON polygon

## Interaction Handlers

### Query vector tiles

- Get features under the mouse pointer   
- Show polygon information on click  

### Popups

- Display a popup on hover  
- Display a popup on click  
- Display a popup  

### Filter data

- Filter symbols by text input  
- Filter features within map view  
- Filter symbols by toggling a list  
- Highlight features containing similar data  
- Highlight features under the mouse pointer  

### Dynamic styling

- Change a map's style  
- Change a layer's color with buttons
- Adjust a layer's opacity  
- Highlight features within a bounding box  
- Data-driven circle coloring by ethnicity  
- Show and hide layers   
- Create a draggable point  
- Update a choropleth layer by zoom level  

### Move around the map

- Create a timeline animation  
- Play map locations as a slideshow  
- Move symbol with the keyboard  
- Center the map on a clicked symbol  
- Fly to a location based on scroll position  
- Fly to a location  
- Slowly fly to a location  
- Fit a map to a bounding box  
- Animate a point  
- Animate a point along a route  
- Game-like controls  

## Controls and overlays

- Disable scroll zoom  
- Change the default position for attribution  
- Toggle interactions  
- Restrict map panning to an area  
- Disable map rotation  
- Display map navigation controls  
- Set pitch and bearing  
- Change a map's language 
- Display a non-interactive map  

## Plugins

### Directions

- Display driving directions  

### Geocoding

- Set a point after Geocoder result  
- Add a geocoder  

### GL Draw

- Show drawn polygon area  

## Misc

- Add a video  
- Add an image  
- Add a WMS source  
- Check for browser support  
- Swipe between maps 
- Measure distances
- Get coordinates of the mouse pointer 
- Add live realtime data

lucaswoj commented:
My high-level concerns are

 1. that the support team loves this page ❤️ 
 2. that we use GL terminology accurately on this page (if it's confusing, let's flag and think about the root problem)
 3. that the inclusion / exclusion of an example in a section is relatively unambiguous 

I'm happy to use @danswick's proposal as a starting point and iterate on it to address concerns 2 & 3.

mourner commented:
Would you be up to merging it as is and iterating as a follow-up? This could take weeks to iterate to a _perfect_ state and it's a huge improvement already in any case.

colleenmcginnis commented:
Sure @mourner - I agree the current iteration is a major improvement. I'll merge now and work with the support and GL JS teams to iterate further.
